### ISSUE TITLE
setFieldValue creates infinite loop

### ISSUE THREAD

**Issue by ivan-sakoman** [formium/formik#3100]
I'm creating CRUD and when trying to create an edit screen and populate fields with values from the server, setFieldValue creates an infinite loop and the app is crashing

logic is working but infinite loop crashes the page

I want to trigger setFieldValue once 

`import { RouteComponentProps,withRouter } from "react-router";
import { Field, Formik, Form, FieldAttributes, useField} from "formik";
import MainWrapper from "MainWrapper";
import * as Yup from 'yup';
import React, { useState } from "react";
import GameRulesService from "GameRulesService";

interface MatchParams
{
    id: string;
}

interface Props extends RouteComponentProps<MatchParams>
{
}

interface FieldsType
{
    headline: string;
}

const AddGameRule = ( props: Props ) =>
{
    const [ items, setItems ] = useState<FieldsType>();
    const editId = props.match.params.id;
    const isAddMode = !editId;

    let initialValues = {
        headline: ''
    }

    const MyTextField: React.FC<FieldAttributes<{}>> = ({placeholder, id, ...props}) =>
    {
        const [ field, meta ] = useField<{}>( props );
        const errorText = meta.error && meta.touched ? meta.error : ''
        return (
            <React.Fragment>
                <Field
                    placeholder={ placeholder }
                    className="input input--base input--text"
                    id={id}
                    { ...field }
                />
                <div className="type--color--error">{ errorText ? errorText : ''}</div>
            </React.Fragment>
        )
    }

    return (
         
        <MainWrapper>
            <Formik
                initialValues={initialValues}
                validationSchema={
                    Yup.object( {
                        headline: Yup.string()
                            .min( 3, 'must be at least 3 characters' )
                            .max( 15, 'must be less than 15 characters' )
                            .required( 'required' )
                    })
                }
                onSubmit={ ( data, { setSubmitting } ) =>
                {
                    setSubmitting( true );
                    
                    const formData = new FormData();

                    const sendValues: any = {
                        headline: data.headline,
                    }

                    for ( const value in sendValues )
                    {
                        formData.append(value, sendValues[value])  
                    }

                    if ( isAddMode )
                    {
                        GameRulesService.addNewGameRule( formData ).then( (e) =>
                        {
                            console.log('response: ', e)
                        } );    
                    } else {
                        GameRulesService.updateSingleGameRule( editId, formData ).then( (e) =>
                        {
                            console.log('response: ', e)
                        } ); 
                    }

                    setSubmitting( false );
                } }
            >
                { ( {  isSubmitting, setFieldValue, errors } ) =>
                {
                    if ( !isAddMode )
                    {
                        GameRulesService.getSingleGameRule( editId ).then( item =>
                        {
                            const sendValues: any = {
                                headline: item.data.headline,
                            }

                            for ( const value in sendValues )
                            {
                                //HERE IS A PROBLEM WHEN I COMMENT THIS LINE UNDER EVERYTHING ELSE IS WORKING 
                                setFieldValue( value, sendValues[value], false )
                            }
                        })

                    }

                    return (
                        
                        <Form
                            id="game-rule"
                        >
                            <div className="field">
                                <label className="field__label" htmlFor="headline">
                                    Headline
                                </label>
                                <MyTextField
                                    name="headline"
                                    id="headline"
                                    placeholder="Enter headline"
                                />
                            </div>
                        </Form>
                    )
                }}
            </Formik>
            {/* <ModalCard /> */ }
        </MainWrapper>
    );
};

export default withRouter(AddGameRule);
`

I need to trigger setFieldValue just once

### Your environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Formik           |   "2.2.6"
| React            |   17.0.1
| TypeScript       |   4.2.2
| Browser          |   Chrome
| npm/Yarn         |   NPM
| Operating System |   Windows

**Comment by ChristopherGillis** [comment]
I'm encountering this bug too. Took me ages to track it down. Downgrading to formik@2.1.5 eliminates the bug.

**Comment by ivan-sakoman** [comment]
I tried but the same problem appears :S

**Comment by johnrom** [comment]
When adding setFieldValue within a child function of Formik, Formik updates during the render, which calls setFieldValue, which updates again, causing renders indefinitely. Formik's actions should only be used during effects (`useEffect`) or the result of a user action like onClick, input, submit, etc.

**Comment by ivan-sakoman** [comment]
i tried to use setFieldValue inside of useEffect, but react and typescript are not happy because i tried to use useEffect inside of formik callback... is there any way to use useEffect inside of Formik callback? 

p.s. this is the reason why I trying to do setFieldValue that way... this is the article which was my guide to make it happen

https://jasonwatmore.com/post/2020/04/17/react-formik-master-details-crud-example

![image](https://user-images.githubusercontent.com/78044930/111641907-e433ae80-87fd-11eb-956e-7391be2ef947.png)

**Comment by johnrom** [comment]
You cannot use an effect within Formik's child function. The child function and render functions are simple functions, it's not a "functional component" in the sense that is required for hooks to be used within it. There are two patterns that can help you:

1. Take your code out of child component and use:

```
const MyFormikChild = (props) => {
  const { isSubmitting, setFieldValue, errors } = useFormikContext();

  useEffect(() =>myEffect(), []);

  return React.Children.only(props.children);
}

const MyForm = () => (
  <Formik {...formikProps}>
    <MyFormikChild>
      <Form /> {/* etc */}
    </MyFormikChild>
  </Formik>
);
```

2. Use `useFormik + FormikProvider`:

```jsx
const MyForm = () => {
  const formik = useFormik(...formikProps);
  useEffect(() => myEffect(), []);

  return <FormikProvider value={formik}>
    <Form /> {/* etc */}
  </FormikProvider>
}
```

**Comment by ivan-sakoman** [comment]
nice, thank you for the answer you helped me a lot :)

**Comment by elham9066** [comment]
از گوشی هوشمند Samsung Galaxy ارسال شده است.
-------- پیام اصلی --------از: "github-actions[bot]" ***@***.***> تاریخ: ۲۰۲۱/۶/۱۸  ۰۴:۳۵  (GMT+03:30) گیرنده: formium/formik ***@***.***> گیرنده کپی: Subscribed ***@***.***> موضوع Re: [formium/formik] setFieldValue creates infinite loop (#3100) 
Closed #3100.

—You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or unsubscribe.
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/formium/formik/issues/3100#event-4907099645",
"url": "https://github.com/formium/formik/issues/3100#event-4907099645",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]
